### DISCUSSION TITLE
How to optimize the impact of binding on program performance?

### DISCUSSION THREAD

**Discussion by kyuranger** [AvaloniaUI/Avalonia#16068] [Q&A]
I have a project (it's a company project, so I cannot provide its source code), in which a usercontrol has about 60 bindings (most of them with converter). These binding functions and styles are irregular, so they cannot be presented in a templated way using listbox or the like.

At the same time, due to the logic of the software function, I cannot split it into several different pages for display in a pagination-like manner.

The problem I'm having now is that every time I load this usercontrol, the UI freezes for about 3 seconds. The strange thing is that this only happens when the program first loads the usercontrol, and not when you close the usercontrol and then reopen it.

After checking other reasons, I found the cause of the problem was binding. I tried to clear all bindings, and the program UI no longer froze. 

The usercontrol was originally bound to a series of static variables, but I later discovered that binding static variables was also a factor affecting its performance. After I changed it to no longer bind static variables but to bind regular variables, the UI still freezes, but the duration of the freeze is reduced to about 1 second. PS: A considerable number of these bindings are two-way bindings.

But what I want to achieve is to completely solve the UI freeze situation. Is there any way to solve the performance problem of a large number of bindings in the same usercontrol?

PS:What I'm actually doing is a settings page similar to the one below:
![image](https://github.com/AvaloniaUI/Avalonia/assets/30720036/9ac38ab9-1fdf-47a2-92f6-57cbac763c57)

**Comment by stevemonaco** [comment]
What platform is this? On desktop, 60 bindings in a "normal" app should not have a serious effect on performance to that degree unless they're generating controls or the converters are doing serious work.

**Reply by kyuranger** [reply]
Windows 10 with Avalonia 11.0.11. There are 11 converters. Most of them only need to control the color of the foreground and whether to display a certain control. They are not complex functions.

**Reply by kyuranger** [reply]
I have an idea to use VirtualizingStackPanel, but VirtualizingStackPanel cannot be used independently, and my needs are not suitable for use with ListBox or the like.

**Comment by maxkatz6** [comment]
If bindings specifically are the bottleneck in your program, please check 11.1 builds. Bindings subsystem was partially reworked there, and it might make a difference for your app.

Also make sure you use only CompiledBindings (can be enabled globally, see docs).

> After checking other reasons, I found the cause of the problem was binding. I tried to clear all bindings, and the program UI no longer froze.

Have you checked that with profiler like dotTrace? It might show where exactly its freezing.

**Reply by kyuranger** [reply]
I have already used CompiledBindings. I haven't tried to test with dotTrace yet, I can try that. I plan to use 11.1 builds after the official release, but it seems that the release is almost ready.

**Reply by maxkatz6** [reply]
Yes, current RC is going to be the same build as stable, unless there are last minute critical changes.

**Reply by rabbitism** [reply]
> Yes, current RC is going to be the same build as stable, unless there are last minute critical changes.

Hope the xaml fix can be included because it breaks third party libraries.

**Reply by kyuranger** [reply]
I tried upgrading to 11.1, but the problem still exists. I tried wrapping my controls with ItemsControl of VirtualizingStackPanel. This way can solve the problem, but it also brings new problems. If the controls in VirtualizingStackPanel are bound to IsVisible, they will not be updated correctly. I'm not sure if this is a bug, but I have tested it in 11.0 and it still exists.

**Reply by stevemonaco** [reply]
`IsVisible` is not a good choice for filtering collections, if that's what you're doing. DynamicData is better for that scenario, but it's a large change.

**Reply by kyuranger** [reply]
![image](https://github.com/AvaloniaUI/Avalonia/assets/30720036/1368e8e7-68ac-4206-9372-1d93c794a3ce)
Actually, I am making a settings page (as shown in the picture above), so my page is a bit like ListBox, but the content/bound elements/styles of each element are different. The problem of IsVisible I mentioned above is that when some functions are not enabled, some settings will not be displayed. So it is not appropriate to use DynamicData.

**Comment by kyuranger** [comment]
I did some tests and found out that the reason why my UI freezes is because there are too many controls. For an application like mine, VirtualizingStackPanel is more suitable. But since I need to dynamically control whether the controls are displayed, this conflicts with the existing design of VirtualizingStackPanel. I looked for it and found that a contributor from Avalonia submitted a commit about controlling the item IsVisible in VirtualizingStackPanel, but it seems that it was rolled back in the end because of too many bugs or other reasons.

So it seems that I can't solve this problem in a short time, and I can only wait for the new commit of controlling the item IsVisible in VirtualizingStackPanel to be released.

**Reply by BAndysc** [reply]
Instead of binding IsVisible, you can just filter the collection you bound in ItemsSource, i.e. if you want to hide - remove from the collection, if you want to display - insert into the collection.
